### ISSUE TITLE
Synthetic keyboard event for Spacebar press has an improper event.key value

### ISSUE THREAD

**Issue by yaycmyk** [facebook/react#5809]
Made a pen for this: http://codepen.io/3lux/pen/PZmJPZ

React is translating a Spacebar press into `" "` instead of `Spacebar`, per the listing on MDN: https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/key

It does look like many characters aren't detected though, so I'm wondering how reliable the use of `event.key` actually is...

**Comment by zpao** [comment]
We actually do normalization so that you aren't getting the differences in platform that are shown in that table (looks like older browsers did return `"Spacebar"` but newer ones return `actual inputted character`, which is `" "` here. Looks like only Firefox and IE/Edge claim support of this draft spec (http://caniuse.com/#feat=keyboardevent-key) so I would say it might not be the most reliable thing but we are doing the normalization to what the spec is expecting (https://github.com/facebook/react/blob/master/src/renderers/dom/client/utils/getEventKey.js).

tldr; as expected, but also cc @syranide in case I was wrong.

**Comment by zpao** [comment]
Here's the relevant section from the current draft of the spec: https://w3c.github.io/uievents/#widl-KeyboardEvent-key

> `key` holds the key value of the key pressed. If the value is has a printed representation, it MUST be a non-empty Unicode character string [...] If the value is a control key which has no printed representation, it MUST be one of the key values defined in the key values set [...] Implementations that are unable to identify a key MUST use the key value `'Unidentified'`.

In this case Spacebar has a printed representation so that's the value.

**Comment by yaycmyk** [comment]
Fair enough, all the normal letter characters I was testing are coming up as `Unidentified` though, which is not what the spec says to do? https://w3c.github.io/uievents/#h-keys-modifiers makes it look like pressing the letter `Q` should set event.key to `"Q"`

**Comment by syranide** [comment]
@zpao Correct, `"spacebar"` is the old draft, it should be `" "` as it is.

@yaycmyk It's impossible to polyfill `key` with characters for `keydown/up` from the virtual key codes reported because there is no way to reliably translate a virtual key code to a character, but we polyfill this where possible (IE mostly IIRC). However we do polyfill characters for `keypress` as it does report character codes. I spent significant time on the `key` polyfill and as far as I am aware (at that time) everything that can be reliably reported is (except for some esoteric keys not worth the added bytes).
